official/projects/lra/moving_average_gated_attention.py

```diff
@@ -0,0 +1,344 @@
+        use_bias=True,
+        kernel_initializer=tf_utils.clone_initializer(gauss_init),
+        bias_initializer=tf_utils.clone_initializer(zero_init),
+        name="v_proj")
+
+    self.mx_proj = tf.keras.layers.Dense(
+        self.zdim + self.hdim + 2 * self.embed_dim,
+        activation=None,
+        use_bias=True,
+        kernel_initializer=tf_utils.clone_initializer(gauss_init),
+        bias_initializer=tf_utils.clone_initializer(zero_init),
+        name="mx_proj")
+
+    self.h_proj = tf.keras.layers.Dense(
+        self.embed_dim,
+        activation=None,
+        use_bias=True,
+        kernel_initializer=tf_utils.clone_initializer(gauss_init),
+        bias_initializer=tf_utils.clone_initializer(zero_init),
+        name="h_proj")
+
+    self._intermediate_dense = tf.keras.layers.Dense(self.inner_dim,
+                                                     use_bias=True)
+
+    self._output_dense = tf.keras.layers.Dense(self.embed_dim, use_bias=True)
+
+    policy = tf.keras.mixed_precision.global_policy()
+    self._intermediate_activation_layer = tf.keras.layers.Activation(
+        self.inner_activation, dtype=policy)
+
+    self.gamma = tf.Variable(gauss_init(shape=[2, self.zdim], dtype=tf.float32),
+                             trainable=True)
+    self.beta = tf.Variable(zero_init(shape=[2, self.zdim], dtype=tf.float32),
+                            trainable=True)
+
+    self.rel_pos_bias = RelativePositionBias(max_positions=self.max_positions)
+
+    super().build(input_shape)
+
+  def get_config(self):
+    config = {
+        "embed_dim":
+            self.embed_dim,
+        "zdim":
+            self.zdim,
+        "hdim":
+            self.hdim,
+        "dropout":
+            self.dropout,
+        "attention_dropout":
+            self.attention_dropout_rate,
+        "kernel_initializer":
+            tf.keras.initializers.serialize(self._kernel_initializer),
+        "bias_initializer":
+            tf.keras.initializers.serialize(self._bias_initializer),
+        "use_bias":
+            self._use_bias,
+        "prenorm":
+            self.prenorm,
+        "max_positions":
+            self.max_positions,
+        "attention_initializer":
+            tf.keras.initializers.serialize(self._attention_initializer),
+        "attention_axes":
+            self._attention_axes,
+        "return_attention_scores":
+            self.return_attention_scores
+    }
+    base_config = super().get_config()
+    return dict(list(base_config.items()) + list(config.items()))
+
+  def softmax_attention(self, q, k):
+    slen = k.shape[1]
+    # C x C
+    if slen is None:
+      slen = 2
+    bias = self.rel_pos_bias(slen)
+
+    # scaled attention
+    q = q * self.scaling
+    # B x K x C x C
+    qk = tf.matmul(q, tf.transpose(k, perm=(0, 2, 1))) + bias
+
+    attn_weights = tf.nn.softmax(qk, axis=-1)
+    return attn_weights
+
+  def call(self, inputs: Any) -> Any:
+    """MEGA encoder block call.
```

Comment:
Please check the formatting. We use double space instead of tab.

Reply:
Thanks for pointing it out! Changed accordingly